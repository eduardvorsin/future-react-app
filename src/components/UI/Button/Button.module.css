.button {
	cursor: pointer;
	background: var(--brand-color-100);
	border-radius: 8px;
	color: var(--secondary-color-100);
	font-size: 16px;
	font-weight: 700;
	line-height: 1.5;
	margin: 0;
	padding: 10px 18px;
	text-align: center;
	border-width: 2px;
	border-style: solid;
	border-color: var(--brand-color-100);
	transform: translateY(0);
	transition:
		background-color var(--transition-time) var(--transition-easing), border-color var(--transition-time) var(--transition-easing),
		color var(--transition-time) var(--transition-easing),
		transform var(--transition-time) var(--transition-easing),
		opacity var(--transition-time) var(--transition-easing);
}

.button:focus {
	border-color: var(--brand-color-100);
	transition: border-color var(--transition-time) var(--transition-easing);
}

.button:disabled {
	cursor: not-allowed;
}

.button:hover:not([disabled]) {
	background-color: var(--secondary-color-100);
	transform: translateY(2px);
	color: var(--brand-color-100);
	transition: background-color var(--transition-time) var(--transition-easing), color var(--transition-time) var(--transition-easing), transform var(--transition-time) var(--transition-easing);
}

.button--icon-button {
	cursor: pointer;
	background-color: transparent;
	color: var(--brand-color-100);
	line-height: 1;
	margin: 0;
	font-size: 0;
	padding: 5px;
	text-align: center;
	border: none;
	display: inline-flex;
	align-items: center;
	justify-content: center;
	transition: color var(--transition-time) var(--transition-easing);
}

.button--icon-button>svg {
	pointer-events: none;
	fill: currentcolor;
	width: 30px;
	height: 30px;
}

.button--icon-button:disabled {
	color: var(--disabled-color-100);
	transition: color var(--transition-time) var(--transition-easing);
	cursor: not-allowed;
}


.button--icon-button:hover:not([disabled]) {
	color: var(--brand-color-200);
	transition: color var(--transition-time) var(--transition-easing);
}

[data-theme='light'] .button:disabled {
	background-color: var(--disabled-color-100);
	color: var(--disabled-color-200);
	border-color: var(--disabled-color-100);
	transition:
		background-color var(--transition-time) var(--transition-easing),
		color var(--transition-time) var(--transition-easing),
		border-color var(--transition-time) var(--transition-easing);
}


[data-theme='dark'] .button:disabled {
	opacity: 0.5;
	transition:
		opacity var(--transition-time) var(--transition-easing);
}