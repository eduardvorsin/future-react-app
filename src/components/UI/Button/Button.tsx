import React from 'react';
import classes from './Button.module.css';

type BaseButtonProps = Partial<React.ButtonHTMLAttributes<HTMLButtonElement>>;
interface ButtonProps extends BaseButtonProps {
  children: React.ReactNode,
  className?: string,
  iconButton?: boolean
  onClick?: React.MouseEventHandler<HTMLButtonElement>
}

const Button: React.FC<ButtonProps> = ({
  children,
  iconButton,
  className,
  onClick,
}) => {
  const buttonClasses = [
    iconButton ? classes['button--icon-button'] : classes.button,
    className,
  ].join(' ');
  return (
    <button
      className={buttonClasses}
      onClick={onClick}
    >
      {children}
    </button>
  );
};

export default Button;
