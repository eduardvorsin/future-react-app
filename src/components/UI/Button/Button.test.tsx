import React, { MouseEvent } from 'react';
import userEvent from '@testing-library/user-event'
import { render, screen } from '@testing-library/react';
import Button from './Button';

describe('Button tests', () => {
  it('is rendered correctly', () => {
    render(<Button>test button</Button>);

    expect(screen.getByRole<HTMLButtonElement>('button', { name: /test button/i })).toBeInTheDocument();
  });

  it('when you click on the button, the mock function should be triggered', async () => {
    const user = userEvent.setup();
    const mockFn = jest.fn<void, [e: MouseEvent]>();
    render(
      <Button
        onClick={mockFn}
      >
        test button
      </Button>,
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button', { name: /test button/i }));

    expect(mockFn).toHaveBeenCalledTimes(1);
  });
  it('when the iconButton prop is true, the button--icon-button class is added', () => {
    render(
      <Button
        iconButton
      >
        test button
      </Button>,
    );

    expect(screen.getByRole<HTMLButtonElement>('button', { name: /test button/i })).toHaveClass('button--icon-button');
  });

  it('is a basic snapshot', () => {
    const mockFn = jest.fn<void, [e: MouseEvent]>();
    render(
      <Button
        className='test-btn'
        onClick={mockFn}
        iconButton
      >
        test button
      </Button>,
    );

    expect(screen.getByRole<HTMLButtonElement>('button', { name: /test button/i })).toMatchSnapshot();
  });
});
