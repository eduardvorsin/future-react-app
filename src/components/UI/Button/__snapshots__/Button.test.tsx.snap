// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tests is a basic snapshot 1`] = `
<button
  class="button--icon-button test-btn"
>
  test button
</button>
`;
